### PR TITLE
Upload build artifacts

### PR DESCRIPTION
**Pull request by iskiselev**
Here I changed travis scripts so that it use travis cache for Nuget/Compiled tests and is able to upload build artifacts (binaries and libraries, compiled tests and translated tests).

To use it:
1. Register [Yandex.Disk](https://disk.yandex.com/invite/?hash=OWD2151Q)
2. Get OAuth key: open [link](https://oauth.yandex.com/authorize?response_type=token&client_id=3dfa760cfe3d4e34bf9f6d8cd98b9de6) when logged in your Yandex.Disk.
4. Create secured environment variable on Travis-CI with name "JSIL_STORAGE_KEY"  and value from step 2.
5. Create "JSIL" folder in Yandex.Disk. Optionally share it (my folder shared r/o with everyone - https://yadi.sk/d/vd1C9YigjN5jt).
6. Yandex.Disk has Web.Dav api, so it can even be attached as network drive in windows (https://yandex.com/support/disk/webdav/webdav-win.xml). There will be 10 Gb of disk space.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -2,25 +2,38 @@ language: csharp
...
   - export NODEJS=$(dirname `nvm which $NODEJS_VERSION`)
   - export FSHARPINSTALLDIR=$PWD/packages/FSharp.Compiler.Tools.4.0.0.1/tools/
-  - ci/download-asset.sh packages .
   - nuget restore JSIL.sln
   - nuget install NUnit.Runners -Version 2.6.4 -OutputDirectory packages 
   - nuget install FSharp.Compiler.Tools -Version 4.0.0.1 -OutputDirectory packages 
-  - ci/upload-asset.sh packages packages &
+
+before_script:
+  - mv /tmp/JSIL_Tests "/tmp/JSIL Tests" || true
```
Comment (kg) [inline .travis.yml]: What's all this about?

File: .travis.yml [hunk 2]
```diff
@@ -2,25 +2,38 @@ language: csharp
...
   - mono ./packages/NUnit.Runners.2.6.4/tools/nunit-console.exe ./bin/Tests.DCE.dll ./bin/SimpleTests.dll ./bin/Tests.dll --run:$TestRun  --exclude:FailsOnMono\|\(FailsOnMonoWhenStubbed+Stubbed\)\|\(FailsOnMonoWhenStubbed+Translated\)
-  - ci/upload-asset.sh compilecache "/tmp/JSIL Tests"
 
+before_cache:
+  - mv "/tmp/JSIL Tests" /tmp/JSIL_Tests  
+  
+after_script:
+  - true && [ ! "$JSIL_STORAGE_TAG" ] && export BUILD_TAG=TRAVIS_${TRAVIS_JOB_NUMBER##*.} || export BUILD_TAG=$JSIL_STORAGE_TAG
+  - export CACHE_PATH=$BUILD_TAG/$TRAVIS_BUILD_NUMBER-$TRAVIS_COMMIT
+  - export UPLOAD_COMMAND="curl -H \"Authorization:OAuth $JSIL_STORAGE_KEY\" -L -T - \"https://webdav.yandex.com/JSIL/_\""
```
Comment (kg) [inline .travis.yml]: So I take it this replaces the service on my box with yandex webdav, and JSIL_STORAGE_KEY needs to be an oauth key?

File: .travis.yml (hunk 1 above)
Comment (iskiselev) [inline .travis.yml]: I found that Travis-CI cache doesn't work if there is space in cached path, so I rename cached compiled tests back here.

Comment (kg) [inline .travis.yml]: Are you sure we want to use travis-ci cache? When I investigated it appeared to not support our scenario (separate cache per ```JSIL_STORAGE_TAG``` value)

File: .travis.yml (hunk 2 above)
Comment (iskiselev) [inline .travis.yml]: Yes, you are correct.
Really, your box was mostly replaced with Travis cache, Yandex is used only for artifacts storage.

File: .travis.yml (hunk 1 above)
Comment (iskiselev) [inline .travis.yml]: I thought about it, but it don't look as critical issue for me. All precompiled tests directories will be merged (maybe not on first run, but not later then 3rd run - you even can restart each sub-build sequentially). After it cache will store precompiled test cases for all tests.

Comment (iskiselev) [inline .travis.yml]: Main benefit here, is that cache accessible (I hope read-only, but have not tested) with pull-request builds, so such builds should also run faster.

**Comment by iskiselev** [comment]
@kg, let's discuss this once again.
Are you have some worries with travis cache? I could disable it then.
I'd like to merge it - otherwise I'm unable to get artifacts of builds that are done on top of master.

It will not hurt if you won't register Yandex.Disk for some time - just leave JSIL_STORAGE_KEY empty (or I can do it for you and send key to write in email).

**Comment by kg** [comment]
Merge has conflicts it looks like, I'm not sure what with

**Comment by iskiselev** [comment]
You'we add some debug output to ci\download-asset.sh and ci\upload-asset.sh, but I've deleted them in this PR.
Rebased on top of current tip.

**Comment by iskiselev** [comment]
@kg, thanks. It will be much easier resolve some strange issues with CI/Mono builds now.

**Pull request merged**